src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,476 @@
+            writer.WriteObjectEnd();
+        }
+
+        protected virtual void WriteCompilationOptionsCore(JsonWriter writer, CompilationOptions options)
+        {
+            // CompilationOption values
+            writer.Write("outputKind", options.OutputKind);
+            writer.Write("moduleName", options.ModuleName);
+            writer.Write("scriptClassName", options.ScriptClassName);
+            writer.Write("mainTypeName", options.MainTypeName);
+            WriteByteArrayValue(writer, "cryptoPublicKey", options.CryptoPublicKey);
+            writer.Write("cryptoKeyFile", options.CryptoKeyFile);
+            writer.Write("delaySign", options.DelaySign);
+            writer.Write("publicSign", options.PublicSign);
+            writer.Write("checkOverflow", options.CheckOverflow);
+            writer.Write("platform", options.Platform);
+            writer.Write("optimizationLevel", options.OptimizationLevel);
+            writer.Write("generalDiagnosticOption", options.GeneralDiagnosticOption);
+            writer.Write("warningLevel", options.WarningLevel);
+            writer.Write("deterministic", options.Deterministic);
+            writer.Write("debugPlusMode", options.DebugPlusMode);
+            writer.Write("referencesSupersedeLowerVersions", options.ReferencesSupersedeLowerVersions);
+            writer.Write("reportSuppressedDiagnostics", options.ReportSuppressedDiagnostics);
+            writer.Write("nullableContextOptions", options.NullableContextOptions);
+
+            writer.WriteKey("specificDiagnosticOptions");
+            writer.WriteArrayStart();
+            foreach (var kvp in options.SpecificDiagnosticOptions)
+            {
+                writer.WriteObjectStart();
+                writer.Write(kvp.Key, kvp.Value);
+                writer.WriteObjectEnd();
+            }
+            writer.WriteArrayEnd();
+
+            if (options.Deterministic)
+            {
+                writer.Write("deterministic", true);
+                writer.WriteNull("localtime");
+            }
+            else
+            {
+                writer.Write("deterministic", false);
+                writer.Write("localtime", options.CurrentLocalTime.ToString(CultureInfo.InvariantCulture));
+            }
+
+            // Values which do not impact build success / failure
+            // - ConcurrentBuild
+            // - MetadataImportOptions:
+            // - Options.Features: deprecated
+            // 
+
+            // Not really options but they can impact compilation so we record the types. For the majority
+            // of compilations this is roughly the equivalent of recording the compiler version but it 
+            // could differ when customers host the compiler via the API.
+            writer.WriteKey("extensions");
+            writer.WriteObjectStart();
+
+            WriteType(writer, "syntaxTreeOptionsProvider", options.SyntaxTreeOptionsProvider?.GetType());
```

Comment:
What determines these values? Concerned that it's going to be difficult to get consistency between the IDE and batch compilations with these.

Reply:
These should be generally the same between the IDE and Compiler. If they are not though we should consider adding an `IgnoreExtensions` flag.